test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs

```diff
@@ -245,6 +256,512 @@ public void PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds()
             }
         }
 
+        /// <summary>
+        /// When pushing a snupkg filename that doesn't exist, show a File Not Found error.
+        /// TODO: bug fixes will come from https://github.com/NuGet/Home/issues/8148
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Snupkg_FilenameDoesNotExist_FileNotFoundError()
+        {
+            // Arrange
+            using (var packageDirectory = TestDirectory.Create())
+            {
+                var nuget = Util.GetNuGetExePath();
+                string snupkgToPush = "nonExistingPackage.snupkg";
+
+                using (var server = CreateAndStartMockV3Server(packageDirectory, out string sourceName))
+                {
+                    // Act
+                    var result = CommandRunner.Run(
+                        nuget,
+                        packageDirectory,
+                        $"push {snupkgToPush} -Source {sourceName} -Timeout 110",
+                        waitForExit: true,
+                        timeOutInMilliseconds: 120000); // 120 seconds
+
+                    // Assert
+
+                    //TODO: Fix push so that this error occurs.
+                    //string expectedFileNotFoundErrorMessage = string.Format(MESSAGE_FILE_DOES_NOT_EXIST, snupkgToPush);
+
+                    Assert.True(0 == result.Item1, "File did not exist and should fail.");
+                    Assert.DoesNotContain(MESSAGE_PACKAGE_PUSHED, result.Item2);
+                    //TODO: Assert.Contains(expectedFileNotFoundErrorMessage, result.Item3);
+                }
+            }
+        }
+
+        /// <summary>
+        /// When pushing a snupkg wildcard where no matching files exist, show a File Not Found error. 
+        /// </summary>
+        [Fact]
+        public void PushCommand_Server_Snupkg_WildcardFindsNothing_FileNotFoundError()
+        {
+            // Arrange
+            using (var packageDirectory = TestDirectory.Create())
+            {
+
+                var nuget = Util.GetNuGetExePath();
+                string snupkgToPush = "*.snupkg";
+
+                using (var server = CreateAndStartMockV3Server(packageDirectory, out string sourceName))
+                {
+                    // Act
+                    var result = CommandRunner.Run(
+                        nuget,
+                        packageDirectory,
+                        $"push {snupkgToPush} -Source {sourceName} -Timeout 110",
+                        waitForExit: true,
+                        timeOutInMilliseconds: 120000); // 120 seconds
+
+                    // Assert
+                    server.Stop();
```

Comment:
Why do you need to stop the server here? Can you rely on the `using` language construct to stop the server?

Reply:
Was following the pattern from original tests. Actually, there's other tests where I could stop it sooner, maybe even dispose it sooner, so thanks for pointing that out.

Reply:
Refactored all tests to dispose the MockServer sooner (before Assertions)